cli/resolver.rs

```diff
@@ -86,81 +93,150 @@ impl CliNodeResolver {
+    permissions: &PermissionsContainer,
+  ) -> Result<Option<NodeResolution>, AnyError> {
+    self.handle_node_resolve_result(self.node_resolver.resolve(
+      specifier,
+      referrer,
+      mode,
+      permissions,
+    ))
+  }
+
   pub fn resolve_req_reference(
     &self,
     req_ref: &NpmPackageReqReference,
     permissions: &PermissionsContainer,
     referrer: &ModuleSpecifier,
-  ) -> Result<ModuleSpecifier, AnyError> {
+    mode: NodeResolutionMode,
+  ) -> Result<NodeResolution, AnyError> {
     let package_folder = self
       .npm_resolver
       .resolve_pkg_folder_from_deno_module_req(req_ref.req(), referrer)?;
-    self
-      .resolve_package_sub_path(
-        &package_folder,
-        req_ref.sub_path(),
-        referrer,
-        permissions,
-      )
-      .with_context(|| format!("Could not resolve '{}'.", req_ref))
+    let maybe_resolution = self.resolve_package_sub_path_from_deno_module(
+      &package_folder,
+      req_ref.sub_path(),
+      referrer,
+      mode,
+      permissions,
+    )?;
+    match maybe_resolution {
+      Some(resolution) => Ok(resolution),
+      None => {
+        if self.npm_resolver.as_byonm().is_some() {
+          let package_json_path = package_folder.join("package.json");
+          if !self.fs.exists_sync(&package_json_path) {
+            return Err(anyhow!(
+              "Could not find '{}'. Deno expects the node_modules/ directory to be up to date. Did you forget to run `npm install`?",
+              package_json_path.display()
+            ));
+          }
+        }
+        Err(anyhow!(
+          "Failed resolving package subpath for '{}' in '{}'.",
+          req_ref,
+          package_folder.display()
+        ))
+      }
+    }
   }
 
-  pub fn resolve_package_sub_path(
+  pub fn resolve_package_sub_path_from_deno_module(
     &self,
     package_folder: &Path,
     sub_path: Option<&str>,
     referrer: &ModuleSpecifier,
+    mode: NodeResolutionMode,
     permissions: &PermissionsContainer,
-  ) -> Result<ModuleSpecifier, AnyError> {
+  ) -> Result<Option<NodeResolution>, AnyError> {
     self.handle_node_resolve_result(
       self.node_resolver.resolve_package_subpath_from_deno_module(
         package_folder,
         sub_path,
         referrer,
-        NodeResolutionMode::Execution,
+        mode,
         permissions,
       ),
     )
   }
 
+  pub fn handle_if_in_node_modules(
+    &self,
+    specifier: ModuleSpecifier,
+  ) -> Result<ModuleSpecifier, AnyError> {
+    // skip canonicalizing if we definitely know it's necessary
+    if specifier.scheme() == "file"
+      && specifier.path().contains("/node_modules/")
+    {
+      // Specifiers in the node_modules directory are canonicalized
+      // so canoncalize then check if it's in the node_modules directory.
+      // If so, check if we need to store this specifier as being a CJS
+      // resolution.
+      let specifier =
+        crate::node::resolve_specifier_into_node_modules(&specifier);
```

Comment:
**thought:** There seems to be a fairly significant amount of cloning happening hereabouts.

`resolve_specifier_into_node_modules` always clones but it seems like it might do extra clones as well. `url_to_node_resolution` below presumably also clones. Inserting into `cjs_resolutions` also clones (this also seems like it might be removable).

It's probably not a real issue but might be worth a glance somewhere down the line if performance in node resolution is found to be somewhat lacking.

Reply:
`resolve_specifier_into_node_modules` seems to conditionally clone? `url_to_node_resolution` does do a to_lowercase to check the extension, which is a clone that could be removed and do a case insensitive comparison. I'm not sure how the clone inserting into cjs_resolutions is removable? Is there a design that could prevent that?